### DISCUSSION TITLE
Questions about the PD controller

### DISCUSSION THREAD

**Discussion by Penguuuuin** [raisimTech/raisimLib#87] [Q&A]
The performance of the provided PD controller is great so I'm wondering how it generates the generalized torque. I've read the document (the theory of the implemented PD controller) but it seems too brief. Users can set the limit of torque, but the limit is not 
reflected in the formula at all. So how the limit works?
I'm trying to calculate the generalized torque by myself. I considered the effect of nonlinear term, M(q)*u, and contact force, but the different between my result and the output of PD controller (read via getGeneralizedForce()) is large. Is there any compensation term that I haven't take into account?
my quadruped robot has no spring and no compliant wire, and joint damping is set to 0

**Comment by jhwangbo** [comment]
The computation of the PD controller is implicit. The issue is that it actually computes the torque limit wrongly because it uses a linear model internally. The model is documented. The whole equation is online too 
https://raisim.com/sections/ArticulatedSystem.html#pd-controller
